Hang in shm/test_ucp_tag_xfer.generic_unexp/0

http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2455/console
```
15:04:31 [       OK ] shm/test_ucp_tag_xfer.generic_unexp/0 (2211 ms)
15:04:31 [ RUN      ] shm/test_ucp_tag_xfer.contig_exp_sync/0
15:04:31 [     INFO ] 0 79x10^1 25x10^2 7x10^3 3x10^4 3x10^5 3x10^6 3x10^7 
16:17:15 make: *** [test_valgrind] Killed
16:17:15 Build step 'Execute shell' marked build as failure
16:17:15 TAP Reports Processing: START
16:17:15 Looking for TAP results report in workspace using pattern: **/*.tap
16:17:15 Saving reports...
```

--- a/test/gtest/common/test_helpers.h
+++ b/test/gtest/common/test_helpers.h
@@ -274,7 +274,7 @@ public:
     }
 
     operator T() const {
-        return m_value;
+        return get();
     }
 
     operator bool() const {
@@ -282,7 +282,7 @@ public:
     }
 
     T get() const {
-        return m_value;
+        return m_initialized ? m_value : NULL;
     }
 
 private:

--- a/test/gtest/ucp/test_ucp_fence.cc
+++ b/test/gtest/ucp/test_ucp_fence.cc
@@ -165,8 +165,8 @@ protected:
 
         ucp_rkey_destroy(rkey);
 
-        sender().disconnect();
-        receiver().disconnect();
+        disconnect(sender());
+        disconnect(receiver());
 
         status = ucp_mem_unmap(receiver().ucph(), memh);
         ASSERT_UCS_OK(status);

--- a/test/gtest/ucp/test_ucp_memheap.cc
+++ b/test/gtest/ucp/test_ucp_memheap.cc
@@ -77,8 +77,8 @@ void test_ucp_memheap::test_nonblocking_implicit_stream_xfer(nonblocking_send_fu
     ucp_rkey_destroy(rkey);
     receiver().flush_worker();
 
-    sender().disconnect();
-    receiver().disconnect();
+    disconnect(sender());
+    disconnect(receiver());
 
     ucp_rkey_buffer_release(rkey_buffer);
     status = ucp_mem_unmap(receiver().ucph(), memh);
@@ -159,8 +159,8 @@ void test_ucp_memheap::test_blocking_xfer(blocking_send_func_t send, size_t alig
     ucp_rkey_destroy(rkey);
     receiver().flush_worker();
 
-    sender().disconnect();
-    receiver().disconnect();
+    disconnect(sender());
+    disconnect(receiver());
 
     status = ucp_mem_unmap(receiver().ucph(), memh);
     ASSERT_UCS_OK(status);

--- a/test/gtest/ucp/test_ucp_tag.cc
+++ b/test/gtest/ucp/test_ucp_tag.cc
@@ -32,14 +32,6 @@ void test_ucp_tag::init()
     dt_gen_finish_count = 0;
 }
 
-void test_ucp_tag::cleanup()
-{
-    sender().flush_worker();
-    receiver().flush_worker();
-    sender().disconnect();
-    ucp_test::cleanup();
-}
-
 void test_ucp_tag::request_init(void *request)
 {
     struct request *req = (struct request *)request;

--- a/test/gtest/ucp/test_ucp_tag.h
+++ b/test/gtest/ucp/test_ucp_tag.h
@@ -37,8 +37,6 @@ protected:
 
     virtual void init();
 
-    virtual void cleanup();
-
     static void request_init(void *request);
 
     static request* request_alloc();

--- a/test/gtest/ucp/test_ucp_tag_match.cc
+++ b/test/gtest/ucp/test_ucp_tag_match.cc
@@ -211,13 +211,6 @@ UCS_TEST_P(test_ucp_tag_match, send2_nb_recv_medium_wildcard, "RNDV_THRESH=-1") 
         request_release(rreq1);
         request_release(rreq2);
     }
-
-    sender().flush_worker();
-    sender2.flush_worker();
-    receiver().flush_worker();
-
-    sender().disconnect();
-    sender2.disconnect();
 }
 
 UCS_TEST_P(test_ucp_tag_match, send_recv_nb_partial_exp_medium) {

--- a/test/gtest/ucp/test_ucp_wireup.cc
+++ b/test/gtest/ucp/test_ucp_wireup.cc
@@ -50,8 +50,6 @@ protected:
 
     void disconnect(ucp_ep_h ep);
 
-    void wait(void *req);
-
     void waitall(std::vector<void*> reqs);
 
 private:
@@ -228,22 +226,6 @@ void test_ucp_wireup::disconnect(ucp_ep_h ep) {
         ASSERT_UCS_OK(UCS_PTR_STATUS(req));
     }
     wait(req);
-}
-
-void test_ucp_wireup::wait(void *req)
-{
-    if (req == NULL) {
-        return;
-    }
-
-    ucs_status_t status;
-    do {
-        progress();
-        ucp_tag_recv_info info;
-        status = ucp_request_test(req, &info);
-    } while (status == UCS_INPROGRESS);
-    ASSERT_UCS_OK(status);
-    ucp_request_release(req);
 }
 
 void test_ucp_wireup::waitall(std::vector<void*> reqs)
@@ -505,7 +487,7 @@ UCS_TEST_P(test_ucp_wireup, disconnect_nb_onesided) {
     std::vector<void*> sreqs;
     send_nb(sender().ep(), 1000, 1000, sreqs);
 
-    void *dreq = ucp_disconnect_nb(sender().revoke_ep());
+    void *dreq = sender().disconnect_nb();
     if (!UCS_PTR_IS_PTR(dreq)) {
         ASSERT_UCS_OK(UCS_PTR_STATUS(dreq));
     }

--- a/test/gtest/ucp/ucp_test.cc
+++ b/test/gtest/ucp/ucp_test.cc
@@ -41,7 +41,7 @@ void ucp_test::cleanup() {
     for (ucs::ptr_vector<entity>::const_iterator iter = entities().begin();
          iter != entities().end(); ++iter)
     {
-        (*iter)->disconnect();
+        disconnect(**iter);
     }
     m_entities.clear();
 }
@@ -86,6 +86,30 @@ void ucp_test::short_progress_loop() const {
         progress();
         usleep(100);
     }
+}
+
+void ucp_test::disconnect(const entity& entity) {
+    void *dreq = entity.disconnect_nb();
+    if (!UCS_PTR_IS_PTR(dreq)) {
+        ASSERT_UCS_OK(UCS_PTR_STATUS(dreq));
+    }
+    wait(dreq);
+}
+
+void ucp_test::wait(void *req)
+{
+    if (req == NULL) {
+        return;
+    }
+
+    ucs_status_t status;
+    do {
+        progress();
+        ucp_tag_recv_info info;
+        status = ucp_request_test(req, &info);
+    } while (status == UCS_INPROGRESS);
+    ASSERT_UCS_OK(status);
+    ucp_request_release(req);
 }
 
 std::vector<ucp_test_param>
@@ -278,6 +302,14 @@ void ucp_test_base::entity::fence() const {
 
 void ucp_test_base::entity::disconnect() {
     m_ep.reset();
+}
+
+void* ucp_test_base::entity::disconnect_nb() const {
+    ucp_ep_h ep = revoke_ep();
+    if (ep == NULL) {
+        return NULL;
+    }
+    return ucp_disconnect_nb(ep);
 }
 
 void ucp_test_base::entity::destroy_worker() {

--- a/test/gtest/ucp/ucp_test.h
+++ b/test/gtest/ucp/ucp_test.h
@@ -34,6 +34,8 @@ public:
 
         void disconnect();
 
+        void* disconnect_nb() const;
+
         void destroy_worker();
 
         ucp_ep_h ep() const;
@@ -95,9 +97,11 @@ public:
 protected:
     virtual void init();
     virtual void cleanup();
-    ucp_test_base::entity* create_entity(bool add_in_front = false);
+    entity* create_entity(bool add_in_front = false);
     void progress() const;
     void short_progress_loop() const;
+    void disconnect(const entity& entity);
+    void wait(void *req);
     static void disable_errors();
     static void restore_errors();
 

